// @docusaurus/plugin-debug/lib/index.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const utils_1 = require("@docusaurus/utils");
const path_1 = __importDefault(require("path"));
function pluginContentPages({ siteConfig: { baseUrl }, }) {
    return {
        name: 'docusaurus-plugin-debug',
        getThemePath() {
            return path_1.default.resolve(__dirname, '../src/theme');
        },
        contentLoaded({ actions: { addRoute } }) {
            addRoute({
                path: utils_1.normalizeUrl([baseUrl, '__docusaurus/debug']),
                component: '@theme/Debug',
                exact: true,
            });
        },
    };
}
exports.default = pluginContentPages;
