// @docusaurus/theme-classic/lib/theme/hooks/useWindowSize.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.windowSizes = void 0;

var _react = require("react");

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const desktopThresholdWidth = 996;
const windowSizes = {
  desktop: 'desktop',
  mobile: 'mobile'
};
exports.windowSizes = windowSizes;

function useWindowSize() {
  const isClient = typeof window !== 'undefined';

  function getSize() {
    if (!isClient) {
      return undefined;
    }

    return window.innerWidth > desktopThresholdWidth ? windowSizes.desktop : windowSizes.mobile;
  }

  const [windowSize, setWindowSize] = (0, _react.useState)(getSize);
  (0, _react.useEffect)(() => {
    if (!isClient) {
      return undefined;
    }

    function handleResize() {
      setWindowSize(getSize());
    }

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);
  return windowSize;
}

var _default = useWindowSize;
exports.default = _default;
