// @docusaurus/theme-classic/lib/theme/MDXComponents/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Link = _interopRequireDefault(require("@docusaurus/Link"));

var _CodeBlock = _interopRequireDefault(require("@theme/CodeBlock"));

var _Heading = _interopRequireDefault(require("@theme/Heading"));

var _stylesModule = _interopRequireDefault(require("./styles.module.css"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const MDXComponents = {
  code: props => {
    const {
      children
    } = props;

    if (typeof children === 'string') {
      if (!children.includes('\n')) {
        return <code {...props} />;
      }

      return <_CodeBlock.default {...props} />;
    }

    return children;
  },
  a: props => <_Link.default {...props} />,
  pre: props => <div className={_stylesModule.default.mdxCodeBlock} {...props} />,
  h1: (0, _Heading.default)('h1'),
  h2: (0, _Heading.default)('h2'),
  h3: (0, _Heading.default)('h3'),
  h4: (0, _Heading.default)('h4'),
  h5: (0, _Heading.default)('h5'),
  h6: (0, _Heading.default)('h6')
};
var _default = MDXComponents;
exports.default = _default;
