// @docusaurus/plugin-content-blog/lib/blogUtils.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getContentPathList = exports.linkify = exports.generateBlogPosts = exports.generateBlogFeed = exports.truncate = void 0;
const fs_extra_1 = __importDefault(require("fs-extra"));
const globby_1 = __importDefault(require("globby"));
const chalk_1 = __importDefault(require("chalk"));
const path_1 = __importDefault(require("path"));
const reading_time_1 = __importDefault(require("reading-time"));
const feed_1 = require("feed");
const utils_1 = require("@docusaurus/utils");
const lodash_1 = require("lodash");
function truncate(fileString, truncateMarker) {
    return fileString.split(truncateMarker, 1).shift();
}
exports.truncate = truncate;
// YYYY-MM-DD-{name}.mdx?
// Prefer named capture, but older Node versions do not support it.
const FILENAME_PATTERN = /^(\d{4}-\d{1,2}-\d{1,2})-?(.*?).mdx?$/;
function toUrl({ date, link }) {
    return `${date
        .toISOString()
        .substring(0, '2019-01-01'.length)
        .replace(/-/g, '/')}/${link}`;
}
async function generateBlogFeed(contentPaths, context, options) {
    if (!options.feedOptions) {
        throw new Error('Invalid options - `feedOptions` is not expected to be null.');
    }
    const { siteConfig } = context;
    const blogPosts = await generateBlogPosts(contentPaths, context, options);
    if (blogPosts == null) {
        return null;
    }
    const { feedOptions, routeBasePath } = options;
    const { url: siteUrl, title, favicon } = siteConfig;
    const blogBaseUrl = utils_1.normalizeUrl([siteUrl, routeBasePath]);
    const updated = (blogPosts[0] && blogPosts[0].metadata.date) ||
        new Date('2015-10-25T16:29:00.000-07:00');
    const feed = new feed_1.Feed({
        id: blogBaseUrl,
        title: feedOptions.title || `${title} Blog`,
        updated,
        language: feedOptions.language,
        link: blogBaseUrl,
        description: feedOptions.description || `${siteConfig.title} Blog`,
        favicon: utils_1.normalizeUrl([siteUrl, favicon]),
        copyright: feedOptions.copyright,
    });
    blogPosts.forEach((post) => {
        const { id, metadata: { title: metadataTitle, permalink, date, description }, } = post;
        feed.addItem({
            title: metadataTitle,
            id,
            link: utils_1.normalizeUrl([siteUrl, permalink]),
            date,
            description,
        });
    });
    return feed;
}
exports.generateBlogFeed = generateBlogFeed;
async function generateBlogPosts(contentPaths, { siteConfig, siteDir }, options) {
    const { include, routeBasePath, truncateMarker, showReadingTime, editUrl, } = options;
    if (!fs_extra_1.default.existsSync(contentPaths.contentPath)) {
        return [];
    }
    const { baseUrl = '' } = siteConfig;
    const blogSourceFiles = await globby_1.default(include, {
        cwd: contentPaths.contentPath,
    });
    const blogPosts = [];
    await Promise.all(blogSourceFiles.map(async (blogSourceFile) => {
        // Lookup in localized folder in priority
        const contentPath = await utils_1.getFolderContainingFile(getContentPathList(contentPaths), blogSourceFile);
        const source = path_1.default.join(contentPath, blogSourceFile);
        const aliasedSource = utils_1.aliasedSitePath(source, siteDir);
        const relativePath = path_1.default.relative(siteDir, source);
        const blogFileName = path_1.default.basename(blogSourceFile);
        const editBlogUrl = utils_1.getEditUrl(relativePath, editUrl);
        const { frontMatter, content, excerpt } = await utils_1.parseMarkdownFile(source);
        if (frontMatter.draft && process.env.NODE_ENV === 'production') {
            return;
        }
        if (frontMatter.id) {
            console.warn(chalk_1.default.yellow(`${blogFileName} - 'id' header option is deprecated. Please use 'slug' option instead.`));
        }
        let date;
        // Extract date and title from filename.
        const match = blogFileName.match(FILENAME_PATTERN);
        let linkName = blogFileName.replace(/\.mdx?$/, '');
        if (match) {
            const [, dateString, name] = match;
            date = new Date(dateString);
            linkName = name;
        }
        // Prefer user-defined date.
        if (frontMatter.date) {
            date = new Date(frontMatter.date);
        }
        // Use file create time for blog.
        date = date || (await fs_extra_1.default.stat(source)).birthtime;
        const slug = frontMatter.slug || (match ? toUrl({ date, link: linkName }) : linkName);
        frontMatter.title = frontMatter.title || linkName;
        blogPosts.push({
            id: frontMatter.slug || frontMatter.title,
            metadata: {
                permalink: utils_1.normalizeUrl([baseUrl, routeBasePath, slug]),
                editUrl: editBlogUrl,
                source: aliasedSource,
                description: frontMatter.description || excerpt,
                date,
                tags: frontMatter.tags,
                title: frontMatter.title,
                readingTime: showReadingTime
                    ? reading_time_1.default(content).minutes
                    : undefined,
                truncated: (truncateMarker === null || truncateMarker === void 0 ? void 0 : truncateMarker.test(content)) || false,
            },
        });
    }));
    blogPosts.sort((a, b) => b.metadata.date.getTime() - a.metadata.date.getTime());
    return blogPosts;
}
exports.generateBlogPosts = generateBlogPosts;
function linkify({ filePath, contentPaths, fileContent, siteDir, blogPosts, onBrokenMarkdownLink, }) {
    // TODO temporary, should consider the file being in localized folder!
    const folderPath = contentPaths.contentPath;
    // TODO perf refactor: do this earlier (once for all md files, not per file)
    const blogPostsBySource = lodash_1.keyBy(blogPosts, (item) => item.metadata.source);
    let fencedBlock = false;
    const lines = fileContent.split('\n').map((line) => {
        if (line.trim().startsWith('```')) {
            fencedBlock = !fencedBlock;
        }
        if (fencedBlock) {
            return line;
        }
        let modifiedLine = line;
        const mdRegex = /(?:(?:\]\()|(?:\]:\s?))(?!https)([^'")\]\s>]+\.mdx?)/g;
        let mdMatch = mdRegex.exec(modifiedLine);
        while (mdMatch !== null) {
            const mdLink = mdMatch[1];
            const aliasedPostSource = `@site/${path_1.default.relative(siteDir, path_1.default.resolve(folderPath, mdLink))}`;
            const blogPost = blogPostsBySource[aliasedPostSource];
            if (blogPost) {
                modifiedLine = modifiedLine.replace(mdLink, blogPost.metadata.permalink);
            }
            else {
                const brokenMarkdownLink = {
                    folderPath,
                    filePath,
                    link: mdLink,
                };
                onBrokenMarkdownLink(brokenMarkdownLink);
            }
            mdMatch = mdRegex.exec(modifiedLine);
        }
        return modifiedLine;
    });
    return lines.join('\n');
}
exports.linkify = linkify;
// Order matters: we look in priority in localized folder
function getContentPathList(contentPaths) {
    return [contentPaths.contentPathLocalized, contentPaths.contentPath];
}
exports.getContentPathList = getContentPathList;
