// @docusaurus/theme-classic/lib/theme/DocItem/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _Head = _interopRequireDefault(require("@docusaurus/Head"));

var _themeCommon = require("@docusaurus/theme-common");

var _useDocusaurusContext = _interopRequireDefault(require("@docusaurus/useDocusaurusContext"));

var _useBaseUrl = _interopRequireDefault(require("@docusaurus/useBaseUrl"));

var _DocPaginator = _interopRequireDefault(require("@theme/DocPaginator"));

var _DocVersionSuggestions = _interopRequireDefault(require("@theme/DocVersionSuggestions"));

var _TOC = _interopRequireDefault(require("@theme/TOC"));

var _IconEdit = _interopRequireDefault(require("@theme/IconEdit"));

var _clsx = _interopRequireDefault(require("clsx"));

var _stylesModule = _interopRequireDefault(require("./styles.module.css"));

var _useDocs = require("@theme/hooks/useDocs");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function DocItem(props) {
  const {
    siteConfig
  } = (0, _useDocusaurusContext.default)();
  const {
    url: siteUrl
  } = siteConfig;
  const {
    content: DocContent
  } = props;
  const {
    metadata,
    frontMatter: {
      image: metaImage,
      keywords,
      hide_title: hideTitle,
      hide_table_of_contents: hideTableOfContents
    }
  } = DocContent;
  const {
    description,
    title,
    permalink,
    editUrl,
    lastUpdatedAt,
    lastUpdatedBy
  } = metadata;
  const {
    pluginId
  } = (0, _useDocs.useActivePlugin)({
    failfast: true
  });
  const versions = (0, _useDocs.useVersions)(pluginId);
  const version = (0, _useDocs.useActiveVersion)(pluginId); // If site is not versioned or only one version is included
  // we don't show the version badge
  // See https://github.com/facebook/docusaurus/issues/3362

  const showVersionBadge = versions.length > 1;
  const metaTitle = (0, _themeCommon.useTitleFormatter)(title);
  const metaImageUrl = (0, _useBaseUrl.default)(metaImage, {
    absolute: true
  });
  return <>
      <_Head.default>
        <title>{metaTitle}</title>
        <meta property="og:title" content={metaTitle} />
        {description && <meta name="description" content={description} />}
        {description && <meta property="og:description" content={description} />}
        {keywords && keywords.length && <meta name="keywords" content={keywords.join(',')} />}
        {metaImage && <meta property="og:image" content={metaImageUrl} />}
        {metaImage && <meta name="twitter:image" content={metaImageUrl} />}
        {metaImage && <meta name="twitter:image:alt" content={`Image for ${title}`} />}
        {permalink && <meta property="og:url" content={siteUrl + permalink} />}
        {permalink && <link rel="canonical" href={siteUrl + permalink} />}
      </_Head.default>

      <div className="row">
        <div className={(0, _clsx.default)('col', {
        [_stylesModule.default.docItemCol]: !hideTableOfContents
      })}>
          <_DocVersionSuggestions.default />
          <div className={_stylesModule.default.docItemContainer}>
            <article>
              {showVersionBadge && <div>
                  <span className="badge badge--secondary">
                    Version: {version.label}
                  </span>
                </div>}
              {!hideTitle && <header>
                  <h1 className={_stylesModule.default.docTitle}>{title}</h1>
                </header>}
              <div className="markdown">
                <DocContent />
              </div>
            </article>
            {(editUrl || lastUpdatedAt || lastUpdatedBy) && <div className="margin-vert--xl">
                <div className="row">
                  <div className="col">
                    {editUrl && <a href={editUrl} target="_blank" rel="noreferrer noopener">
                        <_IconEdit.default />
                        Edit this page
                      </a>}
                  </div>
                  {(lastUpdatedAt || lastUpdatedBy) && <div className="col text--right">
                      <em>
                        <small>
                          Last updated{' '}
                          {lastUpdatedAt && <>
                              on{' '}
                              <time dateTime={new Date(lastUpdatedAt * 1000).toISOString()} className={_stylesModule.default.docLastUpdatedAt}>
                                {new Date(lastUpdatedAt * 1000).toLocaleDateString()}
                              </time>
                              {lastUpdatedBy && ' '}
                            </>}
                          {lastUpdatedBy && <>
                              by <strong>{lastUpdatedBy}</strong>
                            </>}
                          {process.env.NODE_ENV === 'development' && <div>
                              <small>
                                {' '}
                                (Simulated during dev for better perf)
                              </small>
                            </div>}
                        </small>
                      </em>
                    </div>}
                </div>
              </div>}
            <div className="margin-vert--lg">
              <_DocPaginator.default metadata={metadata} />
            </div>
          </div>
        </div>
        {!hideTableOfContents && DocContent.toc && <div className="col col--3">
            <_TOC.default toc={DocContent.toc} />
          </div>}
      </div>
    </>;
}

var _default = DocItem;
exports.default = _default;
