// @docusaurus/core/lib/index.js
"use strict";
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var build_1 = require("./commands/build");
Object.defineProperty(exports, "build", { enumerable: true, get: function () { return build_1.default; } });
var start_1 = require("./commands/start");
Object.defineProperty(exports, "start", { enumerable: true, get: function () { return start_1.default; } });
var swizzle_1 = require("./commands/swizzle");
Object.defineProperty(exports, "swizzle", { enumerable: true, get: function () { return swizzle_1.default; } });
var deploy_1 = require("./commands/deploy");
Object.defineProperty(exports, "deploy", { enumerable: true, get: function () { return deploy_1.default; } });
var external_1 = require("./commands/external");
Object.defineProperty(exports, "externalCommand", { enumerable: true, get: function () { return external_1.default; } });
var serve_1 = require("./commands/serve");
Object.defineProperty(exports, "serve", { enumerable: true, get: function () { return serve_1.default; } });
