// @docusaurus/theme-classic/lib/theme/NavbarItem/DefaultNavbarItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _Link = _interopRequireDefault(require("@docusaurus/Link"));

var _useBaseUrl = _interopRequireDefault(require("@docusaurus/useBaseUrl"));

var _router = require("@docusaurus/router");

var _themeCommon = require("@docusaurus/theme-common");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
function NavLink({
  activeBasePath,
  activeBaseRegex,
  to,
  href,
  label,
  activeClassName = 'navbar__link--active',
  prependBaseUrlToHref,
  ...props
}) {
  // TODO all this seems hacky
  // {to: 'version'} should probably be forbidden, in favor of {to: '/version'}
  const toUrl = (0, _useBaseUrl.default)(to);
  const activeBaseUrl = (0, _useBaseUrl.default)(activeBasePath);
  const normalizedHref = (0, _useBaseUrl.default)(href, {
    forcePrependBaseUrl: true
  });
  return <_Link.default {...href ? {
    target: '_blank',
    rel: 'noopener noreferrer',
    href: prependBaseUrlToHref ? normalizedHref : href
  } : {
    isNavLink: true,
    activeClassName,
    to: toUrl,
    ...(activeBasePath || activeBaseRegex ? {
      isActive: (_match, location) => activeBaseRegex ? new RegExp(activeBaseRegex).test(location.pathname) : location.pathname.startsWith(activeBaseUrl)
    } : null)
  }} {...props}>
      {label}
    </_Link.default>;
}

function NavItemDesktop({
  items,
  position,
  className,
  ...props
}) {
  const dropdownRef = (0, _react.useRef)(null);
  const dropdownMenuRef = (0, _react.useRef)(null);
  const [showDropdown, setShowDropdown] = (0, _react.useState)(false);
  (0, _react.useEffect)(() => {
    const handleClickOutside = event => {
      if (!dropdownRef.current || dropdownRef.current.contains(event.target)) {
        return;
      }

      setShowDropdown(false);
    };

    document.addEventListener('mousedown', handleClickOutside);
    document.addEventListener('touchstart', handleClickOutside);
    return () => {
      document.removeEventListener('mousedown', handleClickOutside);
      document.removeEventListener('touchstart', handleClickOutside);
    };
  }, [dropdownRef]);

  const navLinkClassNames = (extraClassName, isDropdownItem = false) => (0, _clsx.default)({
    'navbar__item navbar__link': !isDropdownItem,
    dropdown__link: isDropdownItem
  }, extraClassName);

  if (!items) {
    return <NavLink className={navLinkClassNames(className)} {...props} />;
  }

  return <div ref={dropdownRef} className={(0, _clsx.default)('navbar__item', 'dropdown', 'dropdown--hoverable', {
    'dropdown--left': position === 'left',
    'dropdown--right': position === 'right',
    'dropdown--show': showDropdown
  })}>
      <NavLink className={navLinkClassNames(className)} {...props} onClick={props.to ? undefined : e => e.preventDefault()} onKeyDown={e => {
      if (e.key === 'Enter') {
        e.preventDefault();
        setShowDropdown(!showDropdown);
      }
    }}>
        {props.label}
      </NavLink>
      <ul ref={dropdownMenuRef} className="dropdown__menu">
        {items.map(({
        className: childItemClassName,
        ...childItemProps
      }, i) => <li key={i}>
            <NavLink onKeyDown={e => {
          if (i === items.length - 1 && e.key === 'Tab') {
            e.preventDefault();
            setShowDropdown(false);
            const nextNavbarItem = dropdownRef.current.nextElementSibling;

            if (nextNavbarItem) {
              nextNavbarItem.focus();
            }
          }
        }} activeClassName="dropdown__link--active" className={navLinkClassNames(childItemClassName, true)} {...childItemProps} />
          </li>)}
      </ul>
    </div>;
}

function NavItemMobile({
  items,
  className,
  position: _position,
  // Need to destructure position from props so that it doesn't get passed on.
  ...props
}) {
  var _menuListRef$current, _menuListRef$current2;

  const menuListRef = (0, _react.useRef)(null);
  const {
    pathname
  } = (0, _router.useLocation)();
  const [collapsed, setCollapsed] = (0, _react.useState)(() => {
    var _items$some;

    return (_items$some = !(items === null || items === void 0 ? void 0 : items.some(item => (0, _themeCommon.isSamePath)(item.to, pathname)))) !== null && _items$some !== void 0 ? _items$some : true;
  });

  const navLinkClassNames = (extraClassName, isSubList = false) => (0, _clsx.default)('menu__link', {
    'menu__link--sublist': isSubList
  }, extraClassName);

  if (!items) {
    return <li className="menu__list-item">
        <NavLink className={navLinkClassNames(className)} {...props} />
      </li>;
  }

  const menuListHeight = ((_menuListRef$current = menuListRef.current) === null || _menuListRef$current === void 0 ? void 0 : _menuListRef$current.scrollHeight) ? `${(_menuListRef$current2 = menuListRef.current) === null || _menuListRef$current2 === void 0 ? void 0 : _menuListRef$current2.scrollHeight}px` : undefined;
  return <li className={(0, _clsx.default)('menu__list-item', {
    'menu__list-item--collapsed': collapsed
  })}>
      <NavLink role="button" className={navLinkClassNames(className, true)} {...props} onClick={() => {
      setCollapsed(state => !state);
    }}>
        {props.label}
      </NavLink>
      <ul className="menu__list" ref={menuListRef} style={{
      height: !collapsed ? menuListHeight : undefined
    }}>
        {items.map(({
        className: childItemClassName,
        ...childItemProps
      }, i) => <li className="menu__list-item" key={i}>
            <NavLink activeClassName="menu__link--active" className={navLinkClassNames(childItemClassName)} {...childItemProps} onClick={props.onClick} />
          </li>)}
      </ul>
    </li>;
}

function DefaultNavbarItem({
  mobile = false,
  ...props
}) {
  const Comp = mobile ? NavItemMobile : NavItemDesktop;
  return <Comp {...props} />;
}

var _default = DefaultNavbarItem;
exports.default = _default;
