// @docusaurus/plugin-content-docs/lib/pluginOptionSchema.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginOptionSchema = exports.DEFAULT_OPTIONS = void 0;
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
const Joi = __importStar(require("@hapi/joi"));
const utils_validation_1 = require("@docusaurus/utils-validation");
const REVERSED_DOCS_HOME_PAGE_ID = '_index';
exports.DEFAULT_OPTIONS = {
    path: 'docs',
    routeBasePath: 'docs',
    homePageId: REVERSED_DOCS_HOME_PAGE_ID,
    include: ['**/*.{md,mdx}'],
    sidebarPath: '',
    docLayoutComponent: '@theme/DocPage',
    docItemComponent: '@theme/DocItem',
    remarkPlugins: [],
    rehypePlugins: [],
    showLastUpdateTime: false,
    showLastUpdateAuthor: false,
    admonitions: {},
    excludeNextVersionDocs: false,
    disableVersioning: false,
};
exports.PluginOptionSchema = Joi.object({
    path: Joi.string().default(exports.DEFAULT_OPTIONS.path),
    editUrl: Joi.string().uri(),
    routeBasePath: Joi.string().allow('').default(exports.DEFAULT_OPTIONS.routeBasePath),
    homePageId: Joi.string().default(exports.DEFAULT_OPTIONS.homePageId),
    include: Joi.array().items(Joi.string()).default(exports.DEFAULT_OPTIONS.include),
    sidebarPath: Joi.string().default(exports.DEFAULT_OPTIONS.sidebarPath),
    docLayoutComponent: Joi.string().default(exports.DEFAULT_OPTIONS.docLayoutComponent),
    docItemComponent: Joi.string().default(exports.DEFAULT_OPTIONS.docItemComponent),
    remarkPlugins: utils_validation_1.RemarkPluginsSchema.default(exports.DEFAULT_OPTIONS.remarkPlugins),
    rehypePlugins: utils_validation_1.RehypePluginsSchema.default(exports.DEFAULT_OPTIONS.rehypePlugins),
    admonitions: utils_validation_1.AdmonitionsSchema.default(exports.DEFAULT_OPTIONS.admonitions),
    showLastUpdateTime: Joi.bool().default(exports.DEFAULT_OPTIONS.showLastUpdateTime),
    showLastUpdateAuthor: Joi.bool().default(exports.DEFAULT_OPTIONS.showLastUpdateAuthor),
    excludeNextVersionDocs: Joi.bool().default(exports.DEFAULT_OPTIONS.excludeNextVersionDocs),
    disableVersioning: Joi.bool().default(exports.DEFAULT_OPTIONS.disableVersioning),
});
